File: phone-app/ios/Runner/background/LocationModifier.swift
import Foundation
import CoreLocation

class LocationModifier {
    private static let EARTH_RADIUS_METERS: Double = 6371E3
    
    public static func perturbLocation(location: CLLocationCoordinate2D) -> CLLocationCoordinate2D {
        let epsilon: Double = Config.getEpsilon()

        
        let sampledBearing = Double.random(in: 0..<2 * Double.pi)
        let sampledProbability: Double = Double.random(in: 0..<1)
        NSLog("I have sampled \(sampledProbability)")
        let sampledLambert: Double = approximateLambertW(givenNumber: (sampledProbability - 1) / M_E)
        NSLog("I have sampled \(sampledLambert)")
        
        let radius: Double = (-1 / epsilon) * (sampledLambert + 1)
        NSLog("Radius is \(radius)")
        
        return addDistanceToLocationAndBearing(location: location, distance: abs(radius), bearing: sampledBearing)
    }
    
    private static func addDistanceToLocationAndBearing(location: CLLocationCoordinate2D, distance: Double, bearing: Double) -> CLLocationCoordinate2D {
        var latitude = location.latitude * Double.pi / 180
        var longitude = location.longitude * Double.pi / 180
        let angularDistance = distance / EARTH_RADIUS_METERS
        let lastLatitude = latitude
        
        latitude = asin(sin(latitude) * cos(angularDistance) + cos(latitude) * sin(angularDistance) * cos(bearing))
        longitude = longitude + atan2(sin(bearing) * sin(angularDistance) * cos(lastLatitude), cos(angularDistance) - sin(latitude) * sin(lastLatitude))
        
        latitude = latitude * 180 / Double.pi
        longitude = longitude * 180 / Double.pi
        NSLog("New location is \(latitude) \(longitude)")
        return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
    
    private static func approximateLambertW(givenNumber: Double) -> Double {
        if (givenNumber == -1 / M_E) {
            return -1
        }
        var last: Double = 1.0
        var iter: Int = 0
        while (true) {
            iter += 1
            let new: Double = last - (last * exp(last) - givenNumber) / (exp(last) * (1 + last))
            if (abs(last - new) < 1e-10) {
                NSLog("iter is \(iter)")
                return round(last * 1e7) / (1e7)
            }
            last = new
        }
    }
}

File: phone-app/android/app/src/main/java/com/example/phone_app/background/LocationModifier.java
package com.example.phone_app.background;

import android.util.Log;
import android.util.Pair;

import com.example.phone_app.config.Config;

public class LocationModifier {
  private final static Double EARTH_RADIUS_METERS = 6371E3;

  public static Pair<Double, Double> perturbLocation(Pair<Double, Double> location) {
    double epsilon = Config.getEpsilon();

    double sampledBearing = Math.random() * 2 * Math.PI;
    double sampledProbability = Math.random();

    Log.i(LocationModifier.class.getName(), "I have sampled " + sampledProbability);

    double sampledLambert = approximateLambertW((sampledProbability - 1) / Math.E);
    Log.i(LocationModifier.class.getName(), "I have sampled " + sampledLambert);

    double radius = Math.abs((-1 / epsilon) * (sampledLambert + 1));
    Log.i(LocationModifier.class.getName(), "Sampled distance is " + radius);

    return addDistanceToLocationAndBearing(location, radius, sampledBearing);
  }

  private static Pair<Double, Double> addDistanceToLocationAndBearing(Pair<Double, Double> location, Double distance, Double bearing) {
    double latitude = Math.toRadians(location.first);
    double longitude = Math.toRadians(location.second);
    double angularDistance = distance / EARTH_RADIUS_METERS;
    double lastLatitude = latitude;


    latitude = Math.asin(Math.sin(latitude) * Math.cos(angularDistance) + Math.cos(latitude) * Math.sin(angularDistance) * Math.cos(bearing));
    longitude = longitude + Math.atan2(Math.sin(bearing) * Math.sin(angularDistance) * Math.cos(lastLatitude), Math.cos(angularDistance) - Math.sin(latitude) * Math.sin(lastLatitude));
    longitude = (longitude + 540) % 360 - 180;
    latitude = Math.toDegrees(latitude);
    longitude = Math.toDegrees(longitude);

    Log.i(LocationModifier.class.getName(), "New location is " + latitude + " " + longitude);

    return new Pair<>(latitude, longitude);
  }

  private static Double approximateLambertW(Double givenProbability) {
    if (givenProbability == -1 / Math.E) {
      return -1.0;
    }

    double last = 1.0;
    while (true) {
      double current = last - (last * Math.exp(last) - givenProbability) / ((last + 1) * Math.exp(last));
      if (Math.abs(current - last) < 1E-10) {
        return Math.round(current * 1E7) / 1E7;
      }
      last = current;
    }
  }
}
